README.md
# cloud_firestore_database_wrapper

## Getting Started

```
instance = FirebaseFirestore.instance;
dataSource = FirestoreDataSource(instance);

final user = User(
          uid: uid,
          name: "Bob",
          photoUrl: "url",
          email: "example@example.com",
          phoneNumber: "1234567890",
          type: 3,
          date: "12/14/2020",
          score: 300,
          userType: []);
await dataSource.create(path, id: user.uid, data: user.toJson());

User user = await dataSource.getSingleByRefId<User>(
            path, originaluser.uid, User.fromJSON);
```

//TODO: Fill in these sections

### Implement FromJson

### Get Single

### Create

### Delete

### Get Collection

### Update

### Get Collection With Params

### Get Collection Stream With Params


CHANGELOG.md
## [1.1.1] - 10/24/2022

- Expose Firestoer parser

## [1.1.0] - 10/24/2022

- Update all methods to use a path based url

## [1.0.2] - 5/15/2021

- Add Base Model for models to inherit from
- Added model generator functionality.

## [1.0.1] - 5/15/2021

- Expose 'getSubcollection' through i_data_source interface

## [1.0.0] - 5/15/2021

- Add 'getSubcollection' functionality
- Package is null safe

## [0.0.3] - 3/29/2021

- Update firestore dependencies

## [0.0.2] - 2/15/2021.

- Update export to include interface

## [0.0.1] - 2/15/2021.

- Initial commit with tests and implementation.


test/generate_model.dart
import 'package:cloud_firestore_database_wrapper/exception/exceptions.dart';
import 'package:cloud_firestore_database_wrapper/src/base_model.dart';
import 'test_models/address.dart';
import 'test_models/user.dart';

Model generateModel<Model extends BaseModel>(
    Map<String, dynamic> json, String id) {
  BaseModel? model;

  if (Model == User) {
    model = User(json);
  } else if (Model == Address) {
    model = Address(json);
  } else {
    Type type = typeOf<Model>();
    throw ModelNotFoundError(
        "Cannont Find Model: " + type.toString(), StackTrace.current);
  }

  return model as Model;
}

Type typeOf<T>() => T;


test/test_models/user.dart
import 'package:cloud_firestore_database_wrapper/src/base_model.dart';

class User extends BaseModel {
  final String? name;
  final String? photoUrl;
  String? email;
  final String? phoneNumber;
  final String? date;
  final int? type;
  final int? score;
  final List? userType;

  User(
    Map<String, dynamic> json,
  )   : name = json["name"],
        email = json["email"],
        phoneNumber = json["phoneNumber"],
        photoUrl = json["photoUrl"],
        type = json["type"],
        score = json["score"],
        date = json["date"],
        userType = json["userType"],
        super(json, json["uid"]);

  Map<String, dynamic> toJSON() => {
        'uid': uid,
        'name': name,
        'email': email,
        'phoneNumber': phoneNumber,
        'photoUrl': photoUrl,
        'type': type,
        'score': score,
        'date': date,
        'userType': userType,
      };

  @override
  String getId() {
    return uid;
  }

  @override
  String getPath() {
    return "users";
  }
}


test/mock_data/mock_user.dart
final users = [
  {
    "name": "Bob",
    "uid": "1",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 1,
    "date": "12/1/2020",
    "userType": [],
    "score": 100
  },
  {
    "name": "Bob",
    "uid": "2",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 1,
    "date": "12/3/2020",
    "userType": [],
    "score": 140
  },
  {
    "name": "Bob",
    "uid": "3",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 2,
    "date": "12/1/2020",
    "userType": [],
    "score": 200
  },
  {
    "name": "Bob",
    "uid": "4",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 3,
    "date": "12/14/2020",
    "userType": [],
    "score": 300
  },
  {
    "name": "Bob",
    "uid": "5",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 1,
    "date": "12/1/2020",
    "userType": [],
    "score": 400
  },
  {
    "name": "Bob",
    "uid": "6",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 2,
    "date": "12/2/2020",
    "userType": [],
    "score": 340
  },
  {
    "name": "Bob",
    "uid": "7",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 3,
    "date": "12/3/2020",
    "userType": [],
    "score": 50
  },
  {
    "name": "Bob",
    "uid": "8",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 2,
    "date": "12/2/2020",
    "userType": [],
    "score": 100
  },
  {
    "name": "Bob",
    "uid": "9",
    "photoUrl": "url",
    "email": "example@example.com",
    "phoneNumber": "1234567890",
    "type": 3,
    "date": "12/1/2020",
    "userType": [],
    "score": 100
  },
] as List<Map<String, dynamic>>;

final addresses = [
  {
    "uid": "1",
    "street": "123 west way",
    "zip": 91234,
    "city": "Los Angeles",
    "state": "CA"
  },
  {
    "uid": "2",
    "street": "443 west way",
    "zip": 91234,
    "city": "Los Angeles",
    "state": "CA"
  },
  {
    "uid": "3",
    "street": "243 west way",
    "zip": 91234,
    "city": "Los Angeles",
    "state": "CA"
  },
  {
    "uid": "4",
    "street": "856 west way",
    "zip": 91234,
    "city": "Los Angeles",
    "state": "CA"
  },
  {
    "uid": "5",
    "street": "645 west way",
    "zip": 91234,
    "city": "Los Angeles",
    "state": "CA"
  },
  {
    "uid": "6",
    "street": "354 west way",
    "zip": 91234,
    "city": "Los Angeles",
    "state": "CA"
  },
  {
    "uid": "7",
    "street": "242 west way",
    "zip": 91234,
    "city": "San Fransisco",
    "state": "CA"
  },
  {
    "uid": "8",
    "street": "434 west way",
    "zip": 91234,
    "city": "San Fransisco",
    "state": "CA"
  },
  {
    "uid": "9",
    "street": "525 west way",
    "zip": 91234,
    "city": "San Fransisco",
    "state": "CA"
  },
] as List<Map<String, dynamic>>;


lib/exception/exceptions.dart
class CloudFirestoreHelperError extends Error {
  final String message;
  final StackTrace stackTrace;
  CloudFirestoreHelperError(this.message, this.stackTrace);

  @override
  String toString() {
    return this.message;
  }
}

class GetSingleDocumentError extends CloudFirestoreHelperError {
  GetSingleDocumentError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class FirestoreReferenceError extends CloudFirestoreHelperError {
  FirestoreReferenceError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class DeleteSingleError extends CloudFirestoreHelperError {
  DeleteSingleError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class UpdateSingleError extends CloudFirestoreHelperError {
  UpdateSingleError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class CreateSingleError extends CloudFirestoreHelperError {
  CreateSingleError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class FailedQueryConstructionError extends CloudFirestoreHelperError {
  FailedQueryConstructionError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class CollectionStreamWithParamsError extends CloudFirestoreHelperError {
  CollectionStreamWithParamsError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class CollectionWithParamsError extends CloudFirestoreHelperError {
  CollectionWithParamsError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class GetCollectionError extends CloudFirestoreHelperError {
  GetCollectionError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class GetCollectionGroupError extends CloudFirestoreHelperError {
  GetCollectionGroupError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class OnlyUseOneError extends CloudFirestoreHelperError {
  OnlyUseOneError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class UnableToCombineError extends CloudFirestoreHelperError {
  UnableToCombineError(StackTrace stackTrace)
      : super(
            'Can\'t use ArrayContainsAny and ArrayContains clauses in the same query',
            stackTrace);
}

class ArrayUseError extends CloudFirestoreHelperError {
  ArrayUseError(StackTrace stackTrace)
      : super(
            'Can\'t use ArrayContainsAny and ArrayContains clauses in the same query',
            stackTrace);
}

class QueryRangeConditionError extends CloudFirestoreHelperError {
  QueryRangeConditionError(String value, String? key, StackTrace stackTrace)
      : super('Duplicate range: $value condition for key: $key', stackTrace);
}

class UnknownQueryError extends CloudFirestoreHelperError {
  UnknownQueryError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class ModelNotFoundError extends CloudFirestoreHelperError {
  ModelNotFoundError(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}

class IsNotDocException extends CloudFirestoreHelperError {
  IsNotDocException(String message, StackTrace stackTrace)
      : super(message, stackTrace);
}


lib/util/firestore_parser.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import 'package:cloud_firestore_database_wrapper/src/base_model.dart';

typedef ModelGenerator<Model> = Function<Model extends BaseModel>(
    Map<String, dynamic> json, String id);

class FirestoreParser<Model extends BaseModel> {
  final ModelGenerator generate;

  FirestoreParser(this.generate);

  List<Model> parse<Model extends BaseModel>(QuerySnapshot querySnapshot) {
    var list = querySnapshot.docs.map((documentSnapshot) {
      var data = generate<Model>(
          documentSnapshot.data() as Map<String, dynamic>,
          documentSnapshot.id) as Model;
      return data;
    }).toList();

    return list;
  }

  Model parseIndividual<Model extends BaseModel>(
      DocumentSnapshot<Map<String, dynamic>> document) {
    var user = generate<Model>(document.data() ?? {}, document.id);
    return user;
  }
}


lib/interfaces/i_data_source.dart
import "dart:async";

import 'package:cloud_firestore_database_wrapper/src/base_model.dart';

enum WhereQueryType {
  isEqualTo,
  isLessThan,
  isLessThanOrEqualTo,
  isGreaterThan,
  isGreaterThanOrEqualTo,
  arrayContains,
  arrayContainsAny,
  whereIn,
  whereNotIn,
  isNull,
  isNotEqualTo,
}

final Set<WhereQueryType> queryRange = {
  WhereQueryType.isLessThan,
  WhereQueryType.isLessThanOrEqualTo,
  WhereQueryType.isGreaterThan,
  WhereQueryType.isGreaterThanOrEqualTo,
};

class QueryType {
  // ignore: prefer_typing_uninitialized_variables
  final id;
  final dynamic value;
  final WhereQueryType? whereQueryType;
  QueryType({this.id, this.value, this.whereQueryType});
}

abstract class IDataSource {
  Future<Model> getSingleById<Model extends BaseModel>(String path);

  Stream<Model> getStreamByID<Model extends BaseModel>(String path);

  Future<List<Model>> getCollectionwithParams<Model extends BaseModel>(
      String path,
      {List<QueryType>? where,
      Map<String, bool>? orderby,
      int? limit,
      String? startAfterID});

  Stream<List<Model>> getCollectionStream<Model extends BaseModel>(
    String path, {
    List<QueryType>? where,
    Map<String, bool>? orderby,
    int? limit,
  });

  Future<String> addDocToCollection<Model extends BaseModel>(
      String path, BaseModel data);

  Future<void> update(String path, BaseModel data);

  delete(String path);
}


lib/src/firestore_data_source.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:cloud_firestore_database_wrapper/exception/exceptions.dart';
import 'package:cloud_firestore_database_wrapper/src/base_model.dart';
import 'package:cloud_firestore_database_wrapper/util/check_query_constructor.dart';
import 'package:cloud_firestore_database_wrapper/util/firestore_parser.dart';

import "dart:async";

import '../interfaces/i_data_source.dart';

// Cloud Firestore
class FirestoreDataSource implements IDataSource {
  FirestoreParser parser;
  FirestoreDataSource(this.parser, this.db);
  final FirebaseFirestore db;

  // Get Single Documents

  @override
  Future<Model> getSingleById<Model extends BaseModel>(
    String path,
  ) async {
    try {
      var pathParam = path.split("/");
      var collection = collectionPath(pathParam);

      if (pathParam.length.isOdd) {
        var document = await collection.doc(pathParam.removeAt(0)).get();
        return parser.parseIndividual<Model>(document);
      } else {
        throw IsNotDocException(
            "Path does not end in a doc id", StackTrace.current);
      }
    } catch (err, s) {
      throw GetSingleDocumentError(err.toString(), s);
    }
  }

  @override
  Stream<Model> getStreamByID<Model extends BaseModel>(
    String path,
  ) {
    var pathParam = path.split("/");
    var collection = collectionPath(pathParam);

    if (pathParam.length.isOdd) {
      var snapshot = collection.doc(pathParam.removeAt(0)).snapshots();
      return snapshot.map((doc) => parser.parseIndividual<Model>(doc));
    } else {
      throw IsNotDocException(
          "Path does not end in a doc id", StackTrace.current);
    }
  }

  // Get collections

  @override
  Future<List<Model>> getCollectionwithParams<Model extends BaseModel>(
      String path,
      {List<QueryType>? where,
      Map<String, bool>? orderby,
      int? limit,
      String? startAfterID}) async {
    var pathParam = path.split("/");

    if (pathParam.length.isEven) {
      throw GetCollectionGroupError(
          "number of ids must be less than paths for subcollection",
          StackTrace.current);
    }

    if (where != null) {
      where = checkQueryConstructor(where);
    }

    try {
      DocumentSnapshot? startAfter;
      if (startAfterID != null) {
        // startAfter = await firestoreRef(path, startAfterID).get();
      }

      var collection = collectionPath(pathParam);

      var collectionReference = queryConstruction(collection,
          where: where, orderby: orderby, limit: limit, startAfter: startAfter);
      var query = await collectionReference.get();
      return parser.parse<Model>(query);
    } catch (err, s) {
      throw CollectionWithParamsError(err.toString(), s);
    }
  }

  @override
  Stream<List<Model>> getCollectionStream<Model extends BaseModel>(
    String path, {
    List<QueryType>? where,
    Map<String, bool>? orderby,
    int? limit,
  }) {
    var pathParam = path.split("/");
    if (pathParam.length.isEven) {
      throw GetCollectionGroupError(
          "number of ids must be less than paths for subcollection",
          StackTrace.current);
    }

    Query query = collectionPath(pathParam);
    if (where != null) {
      where = checkQueryConstructor(where);
    }

    var collection = queryConstruction(
      query,
      where: where,
      orderby: orderby,
      limit: limit,
    );
    Stream<QuerySnapshot> snapshots = collection.snapshots();
    return snapshots.map((snapshot) => parser.parse<Model>(snapshot));
  }

  //Post functions

  @override
  Future<String> addDocToCollection<Model extends BaseModel>(
      String path, BaseModel data) async {
    var pathParam = path.split("/");
    var collection = collectionPath(pathParam);

    if (!pathParam.length.isOdd) {
      var ref = await collection.add(data.toJSON());
      return ref.id;
    } else {
      var id = pathParam.removeAt(0);
      await collection.doc(id).set(data.toJSON());
      return id;
    }
  }

  //Put functions

  @override
  Future<void> update(String path, BaseModel data) async {
    try {
      var pathParam = path.split("/");
      var collection = collectionPath(pathParam);

      if (pathParam.length.isOdd) {
        await collection.doc(pathParam.removeAt(0)).update(data.toJSON());
      } else {
        throw IsNotDocException(
            "Path does not end in a doc id", StackTrace.current);
      }
    } catch (err, s) {
      throw UpdateSingleError(err.toString(), s);
    }
  }

  //Delete functions

  @override
  delete(String path) {
    try {
      var pathParam = path.split("/");
      var collection = collectionPath(pathParam);

      if (pathParam.length.isOdd) {
        collection.doc(pathParam.removeAt(0)).delete();
      } else {
        throw IsNotDocException(
            "Path does not end in a doc id", StackTrace.current);
      }
    } catch (err, s) {
      throw DeleteSingleError(err.toString(), s);
    }
  }

  // Helpers

  Query queryConstruction(Query query,
      {List<QueryType>? where,
      Map<String, bool>? orderby,
      int? limit,
      DocumentSnapshot? startAfter}) {
    try {
      where?.forEach((value) {
        switch (value.whereQueryType) {
          case WhereQueryType.isEqualTo:
            query = query.where(value.id, isEqualTo: value.value);
            break;
          case WhereQueryType.isLessThan:
            query = query.where(value.id, isLessThan: value.value);
            break;
          case WhereQueryType.isLessThanOrEqualTo:
            query = query.where(value.id, isLessThanOrEqualTo: value.value);
            break;
          case WhereQueryType.isGreaterThan:
            query = query.where(value.id, isGreaterThan: value.value);
            break;
          case WhereQueryType.isGreaterThanOrEqualTo:
            query = query.where(value.id, isGreaterThanOrEqualTo: value.value);
            break;
          case WhereQueryType.arrayContains:
            query = query.where(value.id, arrayContains: value.value);
            break;
          case WhereQueryType.arrayContainsAny:
            query = query.where(value.id, arrayContainsAny: value.value);
            break;
          case WhereQueryType.whereIn:
            query = query.where(value.id, whereIn: value.value);
            break;
          case WhereQueryType.whereNotIn:
            query = query.where(value.id, whereNotIn: value.value);
            break;
          case WhereQueryType.isNull:
            query = query.where(value.id, isNull: value.value);
            break;
          case WhereQueryType.isNotEqualTo:
            query = query.where(value.id, isNotEqualTo: value.value);
            break;
          default:
            var type = value.whereQueryType;
            throw UnknownQueryError(
                'Unknown query type $type', StackTrace.current);
        }
      });

      if (orderby != null) {
        orderby.forEach((key, value) {
          query = query.orderBy(key, descending: value);
        });
      }

      if (startAfter != null) {
        query = query.startAfterDocument(startAfter);
      }

      if (limit != null) {
        query = query.limit(limit);
      }

      return query;
    } catch (err, s) {
      throw FailedQueryConstructionError(err.toString(), s);
    }
  }

  CollectionReference<Map<String, dynamic>> collectionPath(
      List<String> pathParam) {
    var collection = db.collection(pathParam.removeAt(0));
    for (var i = 0; i < pathParam.length; i += 2) {
      if (i % 2 == 0 && i < pathParam.length - 1) {
        //collection
        var document = pathParam.removeAt(0);
        var col = pathParam.removeAt(0);
        collection = collection.doc(document).collection(col);
      }
    }

    return collection;
  }
}


lib/util/check_query_constructor.dart
import 'package:cloud_firestore_database_wrapper/exception/exceptions.dart';
import 'package:cloud_firestore_database_wrapper/interfaces/i_data_source.dart';

checkQueryConstructor(List<QueryType> queries) {
  //Validate query list
  Map<String?, QueryType> idmap = {};
  Set<WhereQueryType?> setQuery = {};
  Map<String?, List> rangeMap = {};

  for (var query in queries) {
    if (rangeMap.containsKey(query.id)) {
      rangeMap[query.id]!.add(query.whereQueryType);
    } else {
      rangeMap[query.id] = [query.whereQueryType];
    }

    //You can use at most one in, not-in, or array-contains-any clause per query.
    //You can't combine in , not-in, and array-contains-any in the same query.
    if (setQuery.contains(WhereQueryType.whereIn)) {
      if (setQuery.contains(WhereQueryType.whereNotIn)) {
        throw OnlyUseOneError(
            'Can\'t use whereIn and whereNotIn clauses in the same query',
            StackTrace.current);
      } else if (setQuery.contains(WhereQueryType.arrayContainsAny)) {
        throw OnlyUseOneError(
            'Can\'t use whereIn and ArrayContainsAny clauses in the same query',
            StackTrace.current);
      }
    }
    if (setQuery.contains(WhereQueryType.whereNotIn)) {
      if (setQuery.contains(WhereQueryType.whereIn)) {
        throw OnlyUseOneError(
            'Can\'t use whereIn and WhereNotIn clauses in the same query',
            StackTrace.current);
      } else if (setQuery.contains(WhereQueryType.arrayContainsAny)) {
        throw OnlyUseOneError(
            'Can\'t use ArrayContainsAny and WhereNotIn clauses in the same query',
            StackTrace.current);
      }
    }
    if (setQuery.contains(WhereQueryType.arrayContainsAny)) {
      if (setQuery.contains(WhereQueryType.whereIn)) {
        throw OnlyUseOneError(
            'Can\'t use ArrayContainsAny and WhereIn clauses in the same query',
            StackTrace.current);
      } else if (setQuery.contains(WhereQueryType.whereNotIn)) {
        throw OnlyUseOneError(
            'Can\'t use ArrayContainsAny and WhereNotIn clauses in the same query',
            StackTrace.current);
      }
    }
    if ((query.whereQueryType == WhereQueryType.arrayContains &&
            setQuery.contains(WhereQueryType.arrayContainsAny)) ||
        query.whereQueryType == WhereQueryType.arrayContainsAny &&
            setQuery.contains(WhereQueryType.arrayContains)) {
      //You can't combine array-contains with array-contains-any
      throw UnableToCombineError(StackTrace.current);
    }
    if (query.whereQueryType == WhereQueryType.arrayContains &&
        setQuery.contains(WhereQueryType.arrayContains)) {
      //Use at most one array-contains clause per query.
      throw ArrayUseError(StackTrace.current);
    }
    if (idmap.containsKey(query.id)) {
      //If multiple where clauses on the same id then use where in
      if (idmap[query.id]!.whereQueryType == WhereQueryType.isEqualTo) {
        var oldQ = idmap[query.id]!;
        idmap[query.id] = QueryType(
            id: query.id,
            value: [query.value, oldQ.value],
            whereQueryType: WhereQueryType.whereIn);
      } else if (idmap[query.id]!.whereQueryType == WhereQueryType.whereIn) {
        var list = idmap[query.id]!.value as List;
        list.add(query.value);
      }
    } else {
      idmap[query.id] = query;
      setQuery.add(query.whereQueryType);
    }
  }

  //In a compound query, range (<, <=, >, >=)
  //and not equals (!=) comparisons must all filter on the same field.
  var isRange = false;
  Set keySet = {};
  rangeMap.forEach((key, value) {
    for (var v in value) {
      if (queryRange.contains(v)) {
        if (!isRange) {
          isRange = true;
          break;
        } else {
          if (!keySet.contains(key)) {
            throw QueryRangeConditionError(
                v.toString(), key, StackTrace.current);
          }
        }
      }
    }
    keySet.add(key);
  });

  return idmap.values.toList();
}


example/main.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:cloud_firestore_database_wrapper/src/firestore_data_source.dart';
import 'package:cloud_firestore_database_wrapper/util/firestore_parser.dart';
import '../test/generate_model.dart';
import 'package:flutter/material.dart';
import 'package:firebase_core/firebase_core.dart';
import '../test/mock_data/mock_user.dart' as mockusers;
import '../test/test_models/user.dart';

Future<void> main() async {
  // Pass all uncaught errors to Crashlytics.
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp();

  runApp(
    MyApp(),
  );
}

class MyApp extends StatefulWidget {
  @override
  MyAppState createState() {
    return MyAppState();
  }
}

class MyAppState extends State<MyApp> {
  // The GlobalKey keeps track of the visible state of the list items
  // while they are being animated.
  final GlobalKey<AnimatedListState> listKey = GlobalKey<AnimatedListState>();

  // backing data
  List<User> _data = [];
  final path = 'users';
  final FirebaseFirestore finstance = FirebaseFirestore.instance;
  final FirestoreDataSource dataSource = FirestoreDataSource(
      FirestoreParser(generateModel), FirebaseFirestore.instance);
  num index = 0;

  @override
  Future<void> initState() async {
    for (var user in mockusers.users) {
      var ref = finstance.collection(path).doc(user['uid']);
      ref.set(user);
      index++;
    }
    _data = await dataSource.getCollectionwithParams<User>(path);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        SizedBox(
          height: 300,
          child: AnimatedList(
            // Give the Animated list the global key
            key: listKey,
            initialItemCount: _data.length,
            // Similar to ListView itemBuilder, but AnimatedList has
            // an additional animation parameter.
            itemBuilder: (context, index, animation) {
              // Breaking the row widget out as a method so that we can
              // share it with the _removeSingleItem() method.
              var name = _data[index].name;
              if (name != null) {
                return _buildItem(name, animation);
              }
              return _buildItem("", animation);
            },
          ),
        ),
        ElevatedButton(
          child: Text('Insert item', style: TextStyle(fontSize: 20)),
          onPressed: () {
            _insertSingleItem();
          },
        ),
        ElevatedButton(
          child: Text('Remove item', style: TextStyle(fontSize: 20)),
          onPressed: () {
            _removeSingleItem();
          },
        )
      ],
    );
  }

  // This is the animated row with the Card.
  Widget _buildItem(String name, Animation<double> animation) {
    return SizeTransition(
      sizeFactor: animation,
      child: Card(
        child: ListTile(
          title: Text(
            name,
            style: TextStyle(fontSize: 20),
          ),
        ),
      ),
    );
  }

  void _insertSingleItem() {
    // Arbitrary location for demonstration purposes
    int insertIndex = 2;
    // Add the item to the data list.
    var date = DateTime.now();

    var json = {
      "date": date.toString(),
      "email": "example@example.com",
      "name": "John",
      "photoUrl": "url",
      "phoneNumber": "1234567890",
      "type": 3,
      "userType": [],
      "score": 100
    };
    var user = User(json);
    dataSource.addDocToCollection(path, user);
    _data.insert(insertIndex, user);
  }

  void _removeSingleItem() {
    int removeIndex = 2;
    // Remove item from data list but keep copy to give to the animation.
    User removedItem = _data.removeAt(removeIndex);
    var uid = removedItem.uid;

    dataSource.delete(path + "/" + uid);
    // This builder is just for showing the row while it is still
    // animating away. The item is already gone from the data list.
  }
}
